### PR TITLE
Generate correct value for defaultWorkerPath in worker.config when AOT publishing is enabled

### LINKED ISSUE #1053: [AOT] Update worker.config generation to have correct executable name
The `worker.config` is generated at build time, which contains the executable name. It currently produces incorrect value for `defaultWorkerPath` property (`.dll` extension instead of `.exe`) for an AOT published function app.

### PR DESCRIPTION
**Pull request by kshyju**
resolves #1053

### Pull request checklist

* [x] My changes **do not** require documentation changes
  * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
  * [x] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
  * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] I have added all required tests (Unit tests, E2E tests)

### CODE REVIEW

File: sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets
Comment (kshyju) [inline sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets:70]: There is no single variable present in the `Microsoft.NET.Sdk` targerts file to get this value.

**Review by satvu — APPROVED** [review]

File: sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets
Comment (jviau) [reply sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets:70]: You can use `$(TargetName)$(TargetExt)`

**Review by jviau — CHANGES_REQUESTED** [review]
See comment for using `$(TargetName)$(TargetExt)`

**Comment by kshyju** [comment]
> See comment for using `$(TargetName)$(TargetExt)`

Yea. Switched to the variable in the latest iteration. Thank you!

File: sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets
Comment (kshyju) [reply sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets]: Yea. Switched to the variable in the latest iteration. BTW, My comment was about the absence of the one variable which has the right executable file name value in all different scenarios.

Comment (jviau) [reply sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets]: I meant you do not need to include any of your setting of `_FunctionsWorkerExecutableFileName`. `$(TargetName)$(TargetExt)` will work for all scenarios (.dll, .exe, win-runtime .exe, etc).

File: sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets
Comment (jviau) [inline sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets:77]: This can all be removed and replaced with `$(TargetName)$(TargetExt)`

Comment (kshyju) [reply sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets:77]: No. This produces incorrect value when doing AOT publishing. I just validated it using 
```xml
<Target Name="_PrintFunctionsWorkerExecutableFileName" BeforeTargets="BeforeBuild">
   <Message Text="Printing TargetName TargetExt: $(TargetName)$(TargetExt)" Importance="high" />
</Target>
```
Doing a publish on an app with `<PublishAot>True</PublishAot>`

```bash
shkr  FunctionApp123  ♥ 10:52  dotnet publish -c release -r win-x64 --no-self-contained
MSBuild version 17.8.0+6cdef4241 for .NET
  Determining projects to restore...
  Restored C:\Dev\Temp\FunctionApp123\FunctionApp123\FunctionApp123.csproj (in 552 ms).
C:\Program Files\dotnet\sdk\8.0.100-rc.2.23502.2\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(311,5): message NETSDK1057: You are using a preview version of .NET. S
ee: https://aka.ms/dotnet-support-policy [C:\Dev\Temp\FunctionApp123\FunctionApp123\FunctionApp123.csproj]
  FunctionApp123 -> C:\Dev\Temp\FunctionApp123\FunctionApp123\bin\release\net8.0\win-x64\FunctionApp123.dll
  Printing TargetName TargetExt: FunctionApp123.dll
  Determining projects to restore...
  Restored C:\Users\shkr\AppData\Local\Temp\zesn1q4n.5fb\WorkerExtensions.csproj (in 527 ms).
  WorkerExtensions -> C:\Users\shkr\AppData\Local\Temp\zesn1q4n.5fb\buildout\Microsoft.Azure.Functions.Worker.Extensions.dll
  Generating native code
C:\NUGET_PACKAGES\microsoft.azure.functions.worker.core\1.15.0\lib\net5.0\Microsoft.Azure.Functions.Worker.Core.dll : warning IL2104: Assembly 'Microsoft.Azure.Functions.Worker.Core' produced trim wa
rnings. For more information see https://aka.ms/dotnet-illink/libraries [C:\Dev\Temp\FunctionApp123\FunctionApp123\FunctionApp123.csproj]
C:\NUGET_PACKAGES\microsoft.azure.functions.worker.grpc\1.14.0\lib\net7.0\Microsoft.Azure.Functions.Worker.Grpc.dll : warning IL3053: Assembly 'Microsoft.Azure.Functions.Worker.Grpc' produced AOT ana
lysis warnings. [C:\Dev\Temp\FunctionApp123\FunctionApp123\FunctionApp123.csproj]
C:\NUGET_PACKAGES\microsoft.extensions.configuration.binder\5.0.0\lib\netstandard2.0\Microsoft.Extensions.Configuration.Binder.dll : warning IL2104: Assembly 'Microsoft.Extensions.Configuration.Binde
r' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries [C:\Dev\Temp\FunctionApp123\FunctionApp123\FunctionApp123.csproj]
C:\NUGET_PACKAGES\microsoft.extensions.configuration.binder\5.0.0\lib\netstandard2.0\Microsoft.Extensions.Configuration.Binder.dll : warning IL3053: Assembly 'Microsoft.Extensions.Configuration.Binde
r' produced AOT analysis warnings. [C:\Dev\Temp\FunctionApp123\FunctionApp123\FunctionApp123.csproj]
C:\NUGET_PACKAGES\microsoft.extensions.dependencyinjection\5.0.0\lib\net5.0\Microsoft.Extensions.DependencyInjection.dll : warning IL3053: Assembly 'Microsoft.Extensions.DependencyInjection' produced
 AOT analysis warnings. [C:\Dev\Temp\FunctionApp123\FunctionApp123\FunctionApp123.csproj]
C:\NUGET_PACKAGES\microsoft.extensions.dependencyinjection\5.0.0\lib\net5.0\Microsoft.Extensions.DependencyInjection.dll : warning IL2104: Assembly 'Microsoft.Extensions.DependencyInjection' produced
 trim warnings. For more information see https://aka.ms/dotnet-illink/libraries [C:\Dev\Temp\FunctionApp123\FunctionApp123\FunctionApp123.csproj]
D:\a\_work\1\s\src\DotNetWorker.Core\WorkerInformation.cs(17): warning IL3000: Microsoft.Azure.Functions.Worker.WorkerInformation..cctor(): 'System.Reflection.Assembly.Location.get' always returns an
 empty string for assemblies embedded in a single-file app. If the path to the app directory is needed, consider calling 'System.AppContext.BaseDirectory'. [C:\Dev\Temp\FunctionApp123\FunctionApp123\
FunctionApp123.csproj]
C:\NUGET_PACKAGES\microsoft.azure.functions.worker.grpc\1.14.0\lib\net7.0\Microsoft.Azure.Functions.Worker.Grpc.dll : warning IL2104: Assembly 'Microsoft.Azure.Functions.Worker.Grpc' produced trim wa
rnings. For more information see https://aka.ms/dotnet-illink/libraries [C:\Dev\Temp\FunctionApp123\FunctionApp123\FunctionApp123.csproj]
C:\NUGET_PACKAGES\azure.core\1.10.0\lib\net5.0\Azure.Core.dll : warning IL2104: Assembly 'Azure.Core' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries [C:\Dev\T
emp\FunctionApp123\FunctionApp123\FunctionApp123.csproj]
C:\NUGET_PACKAGES\azure.core\1.10.0\lib\net5.0\Azure.Core.dll : warning IL3053: Assembly 'Azure.Core' produced AOT analysis warnings. [C:\Dev\Temp\FunctionApp123\FunctionApp123\FunctionApp123.csproj]
C:\NUGET_PACKAGES\microsoft.azure.functions.worker.core\1.15.0\lib\net5.0\Microsoft.Azure.Functions.Worker.Core.dll : warning IL3053: Assembly 'Microsoft.Azure.Functions.Worker.Core' produced AOT ana
lysis warnings. [C:\Dev\Temp\FunctionApp123\FunctionApp123\FunctionApp123.csproj]
  FunctionApp123 -> C:\Dev\Temp\FunctionApp123\FunctionApp123\bin\release\net8.0\win-x64\publish\
  Determining projects to restore...
  Restored C:\Users\shkr\AppData\Local\Temp\noszv1bb.2xc\WorkerExtensions.csproj (in 401 ms).
  WorkerExtensions -> C:\Users\shkr\AppData\Local\Temp\noszv1bb.2xc\publishout\Microsoft.Azure.Functions.Worker.Extensions.dll
  WorkerExtensions -> C:\Users\shkr\AppData\Local\Temp\noszv1bb.2xc\publishout\
shkr  FunctionApp123  ♥ 10:53  cd C:\Dev\Temp\FunctionApp123\FunctionApp123\bin\release\net8.0\win-x64\publish\
shkr  publish  ♥ 10:53  dir

    Directory: C:\Dev\Temp\FunctionApp123\FunctionApp123\bin\release\net8.0\win-x64\publish

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d----          10/24/2023 10:53 AM                .azurefunctions
-a---          10/24/2023 10:53 AM            383 extensions.json
-a---          10/24/2023 10:53 AM       10666496 FunctionApp123.exe
-a---          10/24/2023 10:53 AM       82808832 FunctionApp123.pdb
-a---          10/24/2023 10:53 AM              2 functions.metadata
-a---          10/20/2023  8:41 AM            274 host.json
-a---          10/24/2023 10:53 AM            249 worker.config.json
```
You can see what you suggested prints "Printing TargetName TargetExt: FunctionApp123.dll", but our publish output is an executable.

Comment (jviau) [reply sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets:77]: Dug into targets, what is happening is there is a `apphost.exe` that is renamed during copy, so yeah there is no property to check. But I think we can pull this out of an item group. It would be in the `None` item group where identity == `$(AppHostIntermediatePath)`, and the final value we want is the `%(Link)` metadata on that item.

Comment (jviau) [reply sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets:77]:
``` XML
<Target Name="_PrintFunctionsWorkerExecutableFileName" AfterTargets="AfterBuild">
    <ItemGroup>
      <_FunctionsTarget Include="%(None.Link)" Condition="%(Identity) == $(AppHostIntermediatePath)" />
    </ItemGroup>
    <Message Text="Printing TargetName TargetExt: @(_FunctionsTarget)" Importance="high" />
  </Target>
```

Comment (jviau) [reply sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets:77]: Interestingly this exists even with `PublishAot == false`. Wonder why that is?

**Review by jviau — COMMENTED** [review]
I won't block this PR on this, but I think the best approach would be to do something like:

``` XML
    <PropertyGroup>
      <_FunctionsTarget>dotnet $(TargetName)$(TargetExt)</_FunctionsTarget>
      <_FunctionsTarget Condition="('$(SelfContained)' == 'true' OR '$(PublishAot)' == 'true') AND '%(Identity)' == '$(AppHostIntermediatePath)'">%(None.Link)</_FunctionsTarget>
    </PropertyGroup>
```

With that, we can then consolidate all of the different ways to produce `worker.config` with just a single replacement via `_FunctionsTarget`

**Review by jviau — APPROVED** [review]
Approving this if it works for your use case. My SDK refactor PR invalidates a lot of these changes, but I will re-create this behavior there to keep this scenario working.

**Pull request merged**
